p2p/receiver/validator.go

```diff
@@ -31,34 +34,63 @@ func (group ValidatorGroup) Validate(header *p2ppb.Header) (errs error) {
 
 // ValidatorFunc implements Validator and can be used as a adaptor for functions
 // with specific function signature
-type ValidatorFunc func(header *p2ppb.Header) error
+type ValidatorFunc func(envelope *p2ppb.Envelope) error
 
 // Validate passes the arguments to the underlying validator
 // function and returns the results
-func (vf ValidatorFunc) Validate(header *p2ppb.Header) error {
-	return vf(header)
+func (vf ValidatorFunc) Validate(envelope *p2ppb.Envelope) error {
+	return vf(envelope)
 }
 
 func versionValidator() Validator {
-	return ValidatorFunc(func(header *p2ppb.Header) error {
-		if header == nil {
-			return errors.New("nil header")
+	return ValidatorFunc(func(envelope *p2ppb.Envelope) error {
+		if envelope == nil || envelope.Header == nil {
+			return errors.New("nil envelope/header")
 		}
-		if !version.CheckVersion(header.NodeVersion) {
-			return version.IncompatibleVersionError(header.NodeVersion)
+		if !version.CheckVersion(envelope.Header.NodeVersion) {
+			return version.IncompatibleVersionError(envelope.Header.NodeVersion)
 		}
 		return nil
 	})
 }
 
 func networkValidator(networkID uint32) Validator {
-	return ValidatorFunc(func(header *p2ppb.Header) error {
-		if header == nil {
-			return errors.New("nil header")
+	return ValidatorFunc(func(envelope *p2ppb.Envelope) error {
+		if envelope == nil || envelope.Header == nil {
+			return errors.New("nil envelope/header")
+		}
+		if networkID != envelope.Header.NetworkIdentifier {
+			return incompatibleNetworkError(networkID, envelope.Header.NetworkIdentifier)
+		}
+		return nil
+	})
+}
+
+func signatureValidator() Validator {
+	return ValidatorFunc(func(envelope *p2ppb.Envelope) error {
+		if envelope == nil || envelope.Header == nil {
+			return errors.New("nil envelope/header")
+		}
+		envData := &p2ppb.Envelope{
```

Comment:
Any better ideas to deep copy structs?

Reply:
This should help:
https://github.com/golang/protobuf/blob/master/proto/clone.go#L45

Reply:
Thanks!